rust/datafusion/src/execution/physical_plan/parquet.rs

```diff
@@ -0,0 +1,231 @@
+            .filenames
+            .iter()
+            .map(|filename| {
+                Arc::new(ParquetPartition::new(
+                    &filename,
+                    self.projection.clone(),
+                    self.schema.clone(),
+                    self.batch_size,
+                )) as Arc<dyn Partition>
+            })
+            .collect();
+        Ok(partitions)
+    }
+}
+
+struct ParquetPartition {
+    iterator: Arc<Mutex<dyn BatchIterator>>,
+}
+
+impl ParquetPartition {
+    /// Create a new Parquet partition
+    pub fn new(
+        filename: &str,
+        projection: Vec<usize>,
+        schema: Arc<Schema>,
+        batch_size: usize,
+    ) -> Self {
+        // because the parquet implementation is not thread-safe, it is necessary to execute
+        // on a thread and communicate with channels
+        let (request_tx, request_rx): (Sender<()>, Receiver<()>) = unbounded();
+        let (response_tx, response_rx): (
+            Sender<Result<Option<RecordBatch>>>,
+            Receiver<Result<Option<RecordBatch>>>,
+        ) = unbounded();
+
+        let filename = filename.to_string();
+
+        thread::spawn(move || {
+
+            //TODO error handling, remove unwraps
+
+            // open file
+            let file = File::open(&filename).unwrap();
+            let file_reader = Rc::new(SerializedFileReader::new(file).unwrap());
+            let mut arrow_reader = ParquetFileArrowReader::new(file_reader);
+            let mut batch_reader = arrow_reader.get_record_reader_by_columns(projection, batch_size).unwrap();
+
+            while let Ok(_) = request_rx.recv() {
+                match batch_reader.next_batch() {
+                    Ok(Some(batch)) => {
+                        response_tx.send(Ok(Some(batch))).unwrap();
+                    },
+                    Ok(None) => {
+                        response_tx.send(Ok(None)).unwrap();
+                        break;
+                    },
+                    Err(e) => {
+                        response_tx.send(Err(ExecutionError::General(format!("{:?}", e)))).unwrap();
+                        break;
+                    }
+                }
+            }
+        });
+
+        let iterator = Arc::new(Mutex::new(ParquetIterator {
+            schema,
+            request_tx,
+            response_rx,
+        }));
+
+        Self { iterator }
+    }
+}
+
+impl Partition for ParquetPartition {
+    fn execute(&self) -> Result<Arc<Mutex<dyn BatchIterator>>> {
+        Ok(self.iterator.clone())
+    }
+}
+
+struct ParquetIterator {
+    schema: Arc<Schema>,
+    request_tx: Sender<()>,
+    response_rx: Receiver<Result<Option<RecordBatch>>>,
+}
+
+impl BatchIterator for ParquetIterator {
+    fn schema(&self) -> Arc<Schema> {
+        self.schema.clone()
+    }
+
+    fn next(&mut self) -> Result<Option<RecordBatch>> {
+        match self.request_tx.send(()) {
+            Ok(_) => match self.response_rx.recv() {
```

Comment:
Why we need another thread here? This send request, wait response model is also blocked here waiting for IO.

Reply:
We need the threading because the Parquet structs/traits do not implement Sync + Send and cannot be sent between threads.

Reply:
I would much prefer it if we could make Parquet safe to use in multi threaded environments.